// loopback-editor-extension/lib/loopback-editor-extension.js
/**
 * Expose `LoopbackEditorExtension`.
 */

module.exports = LoopbackEditorExtension;

/**
 * Module dependencies.
 */
 
var EventEmitter = require('events').EventEmitter
  , debug = require('debug')('loopback-editor-extension')
  , util = require('util')
  , inherits = util.inherits
  , assert = require('assert')
  , path = require('path')
  , loopback = require('loopback')
  , ejs = require('ejs-locals');
  
/**
 * Create a new `LoopbackEditorExtension` with the given `options`.
 *
 * @param {Object} options
 * @return {LoopbackEditorExtension}
 */

function LoopbackEditorExtension(options) {
  EventEmitter.apply(this, arguments);
  
  // throw an error if args are not supplied
  // assert(typeof options === 'object', 'LoopbackEditorExtension requires an options object');
  
  this.options = options;

  var self = this;

  this.app = loopback();

  this.name = this.options._name;
  this.layout = this.options.layout || '/layouts/layout.ejs';
  
  debug('created with options', options);

  this.app.engine('ejs', ejs);

  // Replace the render function; we want to render the editor layout
  var oldViewDir = this.app.get('views'),
      oldRender = this.app.render,
      app = this.app;

  app.render = function render(view, options, callback) {
    if (typeof options === 'function') {
      callback = options;
      options = {};  
    }

    options.locals = options.locals || {};

    oldRender.call(app, view, options, function(err, body) {
      if (err) return callback(err);
      options._body = body;
      oldRender.call(app, path.join(oldViewDir, self.layout), options, function(err, html) {
        callback(err, html);
      });
    });
  };

  // Set views directory to be relative to the module
  if (this.options._path) {
    this.app.set('views', path.join(path.dirname(this.options._path), "/views"));
    this.app.use(loopback.static(path.join(path.dirname(this.options._path), '/public')));
  }
  
  debug('created with options', options);
}

/**
 * Inherit from `EventEmitter`.
 */

inherits(LoopbackEditorExtension, EventEmitter);

/**
 * Simplified APIs
 */

LoopbackEditorExtension.create =
LoopbackEditorExtension.createLoopbackEditorExtension = function () {
  // add simplified construction / sugar here
  return new LoopbackEditorExtension();
}

/**
 * Methods.
 */
 
LoopbackEditorExtension.prototype.handle = function(url, rootUrl, req, res, next) {
  var oldUrl = req.url;
  req.url = res.viewLocals.relativeUrl = url;
  req.rootUrl = res.viewLocals.rootUrl = rootUrl;
  req.absoluteUrl = res.viewLocals.absoluteUrl = oldUrl;
  var _next = function() {
    req.url = oldUrl;
    next();
  };

  this.app(req, res, _next);
};
