// asteroid-editor-generator/lib/asteroid-editor-generator.js
/**
 * Expose `AsteroidEditorGenerator`.
 */

module.exports = AsteroidEditorGenerator;

/**
 * Module dependencies.
 */
 
var asteroid = require('asteroid')
  , EventEmitter = require('events').EventEmitter
  , debug = require('debug')('asteroid-editor-generator')
  , util = require('util')
  , path = require('path')
  , inherits = util.inherits
  , assert = require('assert')
  , ejs = require('ejs-locals');
  
/**
 * Create a new `AsteroidEditorGenerator` with the given `options`.
 *
 * @param {Object} options
 * @return {AsteroidEditorGenerator}
 */

function AsteroidEditorGenerator(options) {
  EventEmitter.apply(this, arguments);
  
  this.options = options || {};

  this.app = asteroid();

  this.name = this.options._name;
  
  debug('created with options', options);

  this.app.engine('ejs', ejs);

  // Replace the render function; we want to render the editor layout
  var oldViewDir = this.app.get('views'),
      oldRender = this.app.render,
      app = this.app;

  app.render = function render(view, options, callback) {
    if (typeof options === 'function') {
      callback = options;
      options = {};  
    }

    oldRender.call(app, view, options, function(err, body) {
      if (err) return callback(err);

      options._body = body;
      oldRender.call(app, path.join(oldViewDir, '/layouts/editor-layout.ejs'), options, function(err, html) {
        callback(err, html);
      });
    });
  };

  // Set views directory to be relative to the module
  if (this.options._path) {
    this.app.set('views', path.join(path.dirname(this.options._path), "/views"));
  }
}

/**
 * Inherit from `EventEmitter`.
 */

inherits(AsteroidEditorGenerator, EventEmitter);

/**
 * Simplified APIs
 */

AsteroidEditorGenerator.create =
AsteroidEditorGenerator.createAsteroidEditorGenerator = function () {
  // add simplified construction / sugar here
  return new AsteroidEditorGenerator();
}

/**
 * Methods.
 */

AsteroidEditorGenerator.prototype.handle = function(url, req, res, next) {
  var oldUrl = req.url;
  req.url = url;
  var _next = function() {
    req.url = oldUrl;
    next();
  };

  this.app(req, res, _next);
};
